File: src/azure/pull-requests.ts
import { AzureDevOpsClient, pathSegment, withApiVersion } from './client'
import type {
  AzurePullRequest,
  AzurePullRequestChange,
  AzurePullRequestIteration,
  AzureRepository,
  CreatePullRequestInput
} from './types'

export class AzurePullRequestClient {
  constructor(private readonly client: AzureDevOpsClient) {}

  async getRepository(project: string, repository: string, signal?: AbortSignal): Promise<AzureRepository> {
    return (await this.client.request(
      withApiVersion(`/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}`),
      { signal }
    )) as AzureRepository
  }

  async create(
    project: string,
    repository: string,
    request: CreatePullRequestInput,
    signal?: AbortSignal
  ): Promise<AzurePullRequest> {
    return (await this.client.request(
      withApiVersion(
        `/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullrequests`
      ),
      { method: 'POST', body: request, signal }
    )) as AzurePullRequest
  }

  async get(
    project: string,
    repository: string,
    pullRequestId: number,
    signal?: AbortSignal
  ): Promise<AzurePullRequest> {
    return (await this.client.request(
      withApiVersion(
        `/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/${pullRequestId}`
      ),
      { signal }
    )) as AzurePullRequest
  }

  async linkedWorkItemIds(
    project: string,
    repository: string,
    pullRequestId: number,
    signal?: AbortSignal
  ): Promise<number[]> {
    const result = (await this.client.request(
      withApiVersion(
        `/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/${pullRequestId}/workitems`
      ),
      { signal }
    )) as { value: Array<{ id: string | number }> }
    return result.value.map((item) => {
      const id = typeof item.id === 'number' ? item.id : Number(item.id)
      if (!Number.isInteger(id)) throw new Error(`ID de work item inválido: ${String(item.id)}`)
      return id
    })
  }

  async iterations(
    project: string,
    repository: string,
    pullRequestId: number,
    signal?: AbortSignal
  ): Promise<AzurePullRequestIteration[]> {
    const result = (await this.client.request(
      withApiVersion(
        `/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/${pullRequestId}/iterations`,
        '7.0'
      ),
      { signal }
    )) as { value: AzurePullRequestIteration[] }
    return result.value
  }

  async changes(
    project: string,
    repository: string,
    pullRequestId: number,
    iterationId: number,
    signal?: AbortSignal
  ): Promise<AzurePullRequestChange[]> {
    const result = (await this.client.request(
      withApiVersion(
        `/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/${pullRequestId}/iterations/${iterationId}/changes?$top=200`,
        '7.0'
      ),
      { signal }
    )) as { changeEntries: AzurePullRequestChange[] }
    return result.changeEntries.slice(0, 50)
  }
}

File: lib/src/infrastructure/azure/pull_requests.dart
import 'package:better_effect/better_effect.dart';

import '../../app/app_effect.dart';
import 'client.dart';
import 'types.dart';

int? _integerId(Object? value) {
  final number = switch (value) {
    num number => number.toDouble(),
    String text when text.trim().isEmpty => 0,
    String text => double.tryParse(text),
    _ => null,
  };
  if (number == null ||
      !number.isFinite ||
      number != number.truncateToDouble()) {
    return null;
  }
  return number.toInt();
}

abstract interface class AzurePullRequestClient {
  AppEffect<List<AzurePullRequestChange>> changes(
    String project,
    String repository,
    int pullRequestId,
    int iterationId,
  );

  AppEffect<AzurePullRequest> create(
    String project,
    String repository,
    CreatePullRequestInput request,
  );

  AppEffect<AzurePullRequest> get(
    String project,
    String repository,
    int pullRequestId,
  );

  AppEffect<AzureRepository> getRepository(String project, String repository);

  AppEffect<List<AzurePullRequestIteration>> iterations(
    String project,
    String repository,
    int pullRequestId,
  );

  AppEffect<List<int>> linkedWorkItemIds(
    String project,
    String repository,
    int pullRequestId,
  );
}

final class AzurePullRequestClientLive implements AzurePullRequestClient {
  const AzurePullRequestClientLive();

  @override
  AppEffect<List<AzurePullRequestChange>> changes(
    String project,
    String repository,
    int pullRequestId,
    int iterationId,
  ) => Effect.result((use) async {
    final client = use<AzureDevOpsClient>();
    final response = await use.unwrap(
      client.request(
        AzureRequest(
          withApiVersion(
            '/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/$pullRequestId/iterations/$iterationId/changes?${r'$top'}=200',
            '7.0',
          ),
        ),
      ),
    );
    final data = objectMap(response.data);
    final rawItems = data?['changeEntries'];
    if (rawItems is! List) {
      use.fail(
        const AzurePayloadError('Azure não retornou alterações válidas.'),
      );
    }
    return use.unwrap(
      _decode(
        () => rawItems
            .map(objectMap)
            .whereType<Map<String, Object?>>()
            .map(AzurePullRequestChange.fromJson)
            .take(50)
            .toList(growable: false),
      ),
    );
  });

  @override
  AppEffect<AzurePullRequest> create(
    String project,
    String repository,
    CreatePullRequestInput request,
  ) => Effect.result((use) async {
    final client = use<AzureDevOpsClient>();
    final response = await use.unwrap(
      client.request(
        AzureRequest(
          withApiVersion(
            '/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullrequests',
          ),
          method: 'POST',
          body: request.toJson(),
        ),
      ),
    );
    final data = objectMap(response.data);
    if (data == null) {
      use.fail(const AzurePayloadError('Azure não retornou o PR criado.'));
    }
    return use.unwrap(_decode(() => AzurePullRequest.fromJson(data)));
  });

  @override
  AppEffect<AzurePullRequest> get(
    String project,
    String repository,
    int pullRequestId,
  ) => Effect.result((use) async {
    final client = use<AzureDevOpsClient>();
    final response = await use.unwrap(
      client.request(
        AzureRequest(
          withApiVersion(
            '/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/$pullRequestId',
          ),
        ),
      ),
    );
    final data = objectMap(response.data);
    if (data == null) {
      use.fail(const AzurePayloadError('Azure não retornou um PR válido.'));
    }
    return use.unwrap(_decode(() => AzurePullRequest.fromJson(data)));
  });

  @override
  AppEffect<AzureRepository> getRepository(
    String project,
    String repository,
  ) => Effect.result((use) async {
    final client = use<AzureDevOpsClient>();
    final response = await use.unwrap(
      client.request(
        AzureRequest(
          withApiVersion(
            '/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}',
          ),
        ),
      ),
    );
    final data = objectMap(response.data);
    if (data == null) {
      use.fail(
        const AzurePayloadError('Azure não retornou um repositório válido.'),
      );
    }
    return use.unwrap(_decode(() => AzureRepository.fromJson(data)));
  });

  @override
  AppEffect<List<AzurePullRequestIteration>> iterations(
    String project,
    String repository,
    int pullRequestId,
  ) => Effect.result((use) async {
    final client = use<AzureDevOpsClient>();
    final response = await use.unwrap(
      client.request(
        AzureRequest(
          withApiVersion(
            '/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/$pullRequestId/iterations',
            '7.0',
          ),
        ),
      ),
    );
    final data = objectMap(response.data);
    final rawItems = data?['value'];
    if (rawItems is! List) {
      use.fail(
        const AzurePayloadError('Azure não retornou iterações válidas.'),
      );
    }
    return use.unwrap(
      _decode(
        () => rawItems
            .map(objectMap)
            .whereType<Map<String, Object?>>()
            .map(AzurePullRequestIteration.fromJson)
            .toList(growable: false),
      ),
    );
  });

  @override
  AppEffect<List<int>> linkedWorkItemIds(
    String project,
    String repository,
    int pullRequestId,
  ) => Effect.result((use) async {
    final client = use<AzureDevOpsClient>();
    final response = await use.unwrap(
      client.request(
        AzureRequest(
          withApiVersion(
            '/${pathSegment(project)}/_apis/git/repositories/${pathSegment(repository)}/pullRequests/$pullRequestId/workitems',
          ),
        ),
      ),
    );
    final data = objectMap(response.data);
    final rawItems = data?['value'];
    if (rawItems is! List) {
      use.fail(
        const AzurePayloadError(
          'Azure não retornou Work Items vinculados válidos.',
        ),
      );
    }
    final ids = <int>[];
    for (final rawItem in rawItems) {
      final item = objectMap(rawItem);
      final id = _integerId(item?['id']);
      if (item == null || id == null) {
        use.fail(const AzurePayloadError('ID de Work Item inválido.'));
      }
      ids.add(id);
    }
    return ids;
  });
}

AppEffect<T> _decode<T extends Object>(T Function() decode) => Effect.tryAsync(
  decode,
  onError: (error, _) => AzurePayloadError('Resposta Azure inválida: $error'),
);
